### ISSUE TITLE
[Bounty $500] Pole infinity signs diverge from torch: log/log2/log10_bw at x = -0.0, polygamma_bw for n = 0 and even n >= 12

### ISSUE THREAD

**Issue by natesmind** [tenstorrent/tt-metal#55931]
## Summary

Two backward-op composites hard-code the sign of the infinity they emit at their
pole and get it wrong in cases torch defines:

1. `log_bw`, `log2_bw`, `log10_bw` treat `+0.0` and `-0.0` identically. For
   `x = -0.0` they return `+inf * sign(grad)`; torch computes `grad / -0.0` and
   returns `-inf * sign(grad)` — the opposite sign.
2. `polygamma_bw` selects its infinity sign at `x = 0` from a hard-coded list
   (`n == 2 || n == 4 || n == 6 || n == 8 || n == 10`) instead of the parity of
   `n`. For `n = 0` and any even `n >= 12` the sign is inverted relative to
   torch — and relative to tt-metal's own `digamma_bw`, which is the `n = 0`
   case and handles it correctly.

Both are composition-level defects: they reproduce with mathematically ideal
primitives, independent of SFPU kernel accuracy. This is the same divergence
class as the recently approved zero-divisor sign alignment in `xlogy_bw` /
`addcdiv_bw` (#55354), and is disjoint from #55827 (NaN-only, `ttnn.sign`) and
from the catalog in #48240 (no signed-zero or polygamma items).

## 1. log family: `-0.0` produces the wrong-signed infinity

`log2_bw` ([`unary_backward.cpp#L1277-L1299`](https://github.com/tenstorrent/tt-metal/blob/c2d0f71e159d4866c34a6984438fc26d04bcfc73/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1277-L1299), same pattern in `log_bw`
[L840-L858](https://github.com/tenstorrent/tt-metal/blob/c2d0f71e159d4866c34a6984438fc26d04bcfc73/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L840-L858) and `log10_bw`
[L1102-L1124](https://github.com/tenstorrent/tt-metal/blob/c2d0f71e159d4866c34a6984438fc26d04bcfc73/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1102-L1124), main @ c2d0f71e):

```cpp
Tensor t_inf = where(ttnn::ltz(grad, ...), -inf, +inf, ...);   // sign from grad only
...
grad_a = where(
    ttnn::logical_and(ttnn::eqz(input, ...), ttnn::eqz(grad, ...)), NaN,
    where(ttnn::eqz(input, ...), t_inf, grad_a, ...), ...);
```

`eqz` is true for both zeros, and `t_inf` looks only at `grad`, so the sign of
the zero is discarded. Torch's reference is `grad / (x * ln2)`, and IEEE-754
division by `-0.0` flips the sign:

| x    | grad | ttnn  | torch |
|------|------|-------|-------|
| -0.0 | +1   | +inf  | -inf  |
| -0.0 | -1   | -inf  | +inf  |
| +0.0 | +1   | +inf  | +inf (agrees) |

Negative zeros are not exotic inputs: any negative product or division that
underflows in bf16/fp32 produces `-0.0`, which then feeds a wrong-signed
infinite gradient into the graph.

### Proposed fix

Make the pole sign respect the zero's sign bit with existing primitives
(`ttnn::signbit` is already used this way in `acosh_bw`):

```cpp
// zero_sign = -1 for -0.0, +1 for +0.0
Tensor zero_sign = ttnn::where(ttnn::signbit(input, ...), -1.0f, 1.0f, ...);
Tensor t_inf = ttnn::multiply(
    ttnn::multiply(ttnn::sign(grad, ...), zero_sign, ...), +inf, ...);
```

(or equivalently drop the override and let an IEEE-correct reciprocal produce
the sign, keeping the override only for dtypes whose reciprocal kernel does not
honour signed zero).

`grad == 0` at the pole stays NaN, which already matches torch (`0 / ±0`).

## 2. polygamma_bw: parity by lookup table

`polygamma_bw` ([`unary_backward.cpp#L1465-L1499`](https://github.com/tenstorrent/tt-metal/blob/c2d0f71e159d4866c34a6984438fc26d04bcfc73/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1465-L1499)):

```cpp
float pos_neg = 1.0f;
if (n == 2 || n == 4 || n == 6 || n == 8 || n == 10) {
    pos_neg = -1.0f;
}
...
// at input == 0:  gtz(grad) -> (-inf * pos_neg),  ltz(grad) -> (+inf * pos_neg)
```

The multiplier at the pole is `polygamma(n+1, x)`, and as `x -> 0+`,
`polygamma(k, x) ~ (-1)^(k+1) * k! * x^(-k-1)`, so the correct sign for
`grad > 0` is `(-1)^n * inf`: `+inf` for every even `n`, `-inf` for every odd
`n`. The list only covers even `n` in [2, 10]:

| n  | ttnn @ x=0, grad>0 | torch | |
|----|--------------------|-------|---|
| 0  | -inf | +inf | wrong |
| 1  | -inf | -inf | ok    |
| 2  | +inf | +inf | ok    |
| 10 | +inf | +inf | ok    |
| 12 | -inf | +inf | wrong |

`n = 0` is not hypothetical: `polygamma(0, x)` is `digamma`, and
`digamma_bw` ([L1436-L1463](https://github.com/tenstorrent/tt-metal/blob/c2d0f71e159d4866c34a6984438fc26d04bcfc73/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1436-L1463)) returns `+inf` for `grad > 0` at `x = 0` — so
`ttnn.polygamma_bw(grad, x, 0)` and `ttnn.digamma_bw(grad, x)` disagree with
each other on the same mathematical function.

### Proposed fix

```cpp
float pos_neg = (n % 2 == 0) ? -1.0f : 1.0f;
```

## Reproduction (host-only, no card required)

Source-level replay of the compositions with ideal primitives against torch
autograd on float32 (`pip install torch`):

```python
import torch
INF, NAN = float("inf"), float("nan")
t = lambda v: torch.tensor(v, dtype=torch.float32)

def tt_log2_bw(grad, x):                      # unary_backward.cpp L1277-L1299
    t_inf = torch.where(grad < 0, t(-INF), t(INF))
    grad_a = grad * torch.reciprocal(x * 0.6931471805599453)
    return torch.where((x == 0) & (grad == 0), t(NAN),
                       torch.where(x == 0, t_inf, grad_a))

def torch_ref(fn, grad, x):
    x = x.clone().requires_grad_(True)
    (g,) = torch.autograd.grad(fn(x), x, grad_outputs=grad)
    return g

xs, gs = t([-0.0, -0.0, 0.0]), t([1.0, -1.0, 1.0])
print("ttnn :", tt_log2_bw(gs, xs))    # [ inf, -inf, inf]
print("torch:", torch_ref(torch.log2, gs, xs))  # [-inf,  inf, inf]

for n in (0, 1, 2, 12):
    pos_neg = -1.0 if n in (2, 4, 6, 8, 10) else 1.0
    ttnn_val = -INF * pos_neg                      # value forced for grad > 0
    x = torch.zeros(1, dtype=torch.float32).requires_grad_(True)
    (ref,) = torch.autograd.grad(torch.polygamma(n, x), x,
                                 grad_outputs=torch.ones(1))
    print(f"polygamma_bw n={n}: ttnn {ttnn_val:+.0f}  torch {ref.item():+.0f}")
# n=0: ttnn -inf torch +inf   n=12: ttnn -inf torch +inf
```

Caveat: I do not have a card; the tables above are source-level replays of the
host composition (the `where` cascades quoted), which is where both defects
live. Device validation would go through the existing backward sweeps with
`-0.0` and `n in {0, 12}` probes added.

## Scope of the fix

- `log_bw`, `log2_bw`, `log10_bw`: sign-aware `t_inf` (shared helper), ~15
  lines total, plus `-0.0` cases in the unary backward sweeps.
- `polygamma_bw`: one-line parity fix plus `n = 0` / `n = 12` pole cases, and a
  consistency assertion against `digamma_bw`.

Note: this bounty is suggested under the community-suggested-bounty process
(cf. #53787 / #54551); proposed tier: easy-medium. I'd like to take the
implementation if approved — happy to be assigned.

**Comment by mateusznowakTT** [comment]
@natesmind : For bounty proposals please first submit it to bounties@tenstorrent.com for review.
